interfaces/builtin/gpio_chardev.go

```diff
@@ -0,0 +1,212 @@
+		return fmt.Errorf("gpio-chardev interface requires the %q flag to be set", flag)
+	}
+	return nil
+}
+
+func (iface *gpioChardevInterface) SystemdConnectedSlot(spec *systemd.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	var sourceChip []string
+	if err := slot.Attr("source-chip", &sourceChip); err != nil {
+		return err
+	}
+	var lines string
+	if err := slot.Attr("lines", &lines); err != nil {
+		return err
+	}
+	snapName := slot.Snap().InstanceName()
+	slotName := slot.Name()
+
+	serviceSuffix := fmt.Sprintf("gpio-chardev-%s", slotName)
+	service := &systemd.Service{
+		Type:            "oneshot",
+		RemainAfterExit: true,
+		// snap-gpio-helper export-chardev "<chip-labels>" "<lines>" "<gadget>" "<slot-name>"
+		ExecStart: fmt.Sprintf("%s/snap-gpio-helper export-chardev %q %q %q %q",
+			dirs.DistroLibExecDir, strings.Join(sourceChip, ","), lines, snapName, slotName),
+		// snap-gpio-helper unexport-chardev "<chip-labels>" "<lines>" "<gadget>" "<slot-name>"
+		ExecStop: fmt.Sprintf("%s/snap-gpio-helper unexport-chardev %q %q %q %q",
+			dirs.DistroLibExecDir, strings.Join(sourceChip, ","), lines, snapName, slotName),
+		// snapd.gpio-chardev-setup.target is used for synchronization of
+		// app services and virtual device setup during boot.
+		WantedBy: "snapd.gpio-chardev-setup.target",
+		Before:   "snapd.gpio-chardev-setup.target",
+	}
+	return spec.AddService(serviceSuffix, service)
+}
+
+func (iface *gpioChardevInterface) SystemdConnectedPlug(spec *systemd.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	slotName := slot.Name()
+	slotSnapName := slot.Snap().InstanceName()
+	plugName := slot.Name()
+	plugSnapName := plug.Snap().InstanceName()
+
+	target := fmt.Sprintf("/dev/snap/gpio-chardev/%s/%s", slotSnapName, slotName)
+	symlink := fmt.Sprintf("/dev/snap/gpio-chardev/%s/%s", plugSnapName, plugName)
+
+	// Create symlink pointing to exported virtual slot device.
+	serviceSuffix := fmt.Sprintf("gpio-chardev-%s", plugName)
+	service := &systemd.Service{
+		Type:            "oneshot",
+		RemainAfterExit: true,
+		ExecStart:       fmt.Sprintf("/bin/sh -c 'mkdir -p %q && ln -s %q %q'", filepath.Dir(symlink), target, symlink),
+		ExecStop:        fmt.Sprintf("/bin/sh -c 'rm -f %q'", symlink),
+		WantedBy:        "snapd.gpio-chardev-setup.target",
+		Before:          "snapd.gpio-chardev-setup.target",
+	}
+	return spec.AddService(serviceSuffix, service)
+}
+
+func (iface *gpioChardevInterface) AppArmorConnectedPlug(spec *apparmor.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	slotSnapName := slot.Snap().InstanceName()
+	plugSnapName := plug.Snap().InstanceName()
+	snippet := "# Allow access to exported gpio chardev lines\n"
+	// Allow access to exported virtual slot device.
+	snippet += fmt.Sprintf("/dev/snap/gpio-chardev/%s/%s rwk,\n", slotSnapName, slot.Name())
```

Comment:
We are computing this many times in the file. Can you use a helper that has the path exactly once.

Reply:
Good point, I already did that in the helper command PR as a follow up

https://github.com/canonical/snapd/pull/15212/files#diff-10f4c2a34c9c6fe86ebdebad743920f2190b6af510ee25458ec839a7369d9e05